flink-connector-oceanbase-cdc/src/main/java/com/ververica/cdc/connectors/oceanbase/source/OceanBaseRichParallelSourceFunction.java

```diff
@@ -0,0 +1,518 @@
+import org.apache.flink.util.FlinkRuntimeException;
+
+import com.mysql.jdbc.ResultSetMetaData;
+import com.oceanbase.clogproxy.client.LogProxyClient;
+import com.oceanbase.clogproxy.client.config.ObReaderConfig;
+import com.oceanbase.clogproxy.client.exception.LogProxyClientException;
+import com.oceanbase.clogproxy.client.listener.RecordListener;
+import com.oceanbase.oms.logmessage.DataMessage;
+import com.oceanbase.oms.logmessage.LogMessage;
+import com.ververica.cdc.debezium.DebeziumDeserializationSchema;
+import io.debezium.relational.TableSchema;
+import org.apache.kafka.connect.data.Schema;
+import org.apache.kafka.connect.data.Struct;
+import org.apache.kafka.connect.source.SourceRecord;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.sql.Connection;
+import java.sql.DriverManager;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.sql.Statement;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.stream.Collectors;
+
+/**
+ * The source implementation for OceanBase that read snapshot events first and then read the change
+ * event.
+ *
+ * @param <T> The type created by the deserializer.
+ */
+public class OceanBaseRichParallelSourceFunction<T> extends RichSourceFunction<T>
+        implements CheckpointListener, CheckpointedFunction, ResultTypeQueryable<T> {
+
+    private static final long serialVersionUID = 2844054619864617340L;
+
+    private static final Logger LOG =
+            LoggerFactory.getLogger(OceanBaseRichParallelSourceFunction.class);
+
+    private final boolean snapshot;
+    private final String username;
+    private final String password;
+    private final String tenantName;
+    private final String databaseName;
+    private final String tableName;
+    private final String jdbcUrl;
+    private final String rsList;
+    private final String logProxyHost;
+    private final int logProxyPort;
+    private final long startTimestamp;
+    private final DebeziumDeserializationSchema<T> deserializer;
+
+    private final AtomicBoolean snapshotCompleted = new AtomicBoolean(false);
+    private final List<LogMessage> logMessageBuffer = new LinkedList<>();
+
+    private transient Map<String, TableSchema> tableSchemaMap;
+    private transient volatile long resolvedTimestamp;
+    private transient Connection snapshotConnection;
+    private transient LogProxyClient logProxyClient;
+    private transient ListState<Long> offsetState;
+    private transient OutputCollector<T> outputCollector;
+
+    public OceanBaseRichParallelSourceFunction(
+            boolean snapshot,
+            String username,
+            String password,
+            String tenantName,
+            String databaseName,
+            String tableName,
```

Comment:
Currently we can only capture one table, I think capture multiple table is a common usercase for example user's business table is sharding table like (user_id_table1, user_id_table_2,...)

Reply:
Updated.